src/java/org/apache/cassandra/service/accord/CommandsForRanges.java

```diff
@@ -0,0 +1,505 @@
+            for (Range range : summary.ranges)
+            {
+                rangeToTxn.add(Interval.create(normalize(range.start(), range.startInclusive(), true),
+                                               normalize(range.end(), range.endInclusive(), false),
+                                               summary));
+            }
+        }
+
+        public T putAll(CommandsForRanges other)
+        {
+            for (TxnId id : other.localCommands)
+            {
+                TxnType thisType = type(id);
+                TxnType otherType = other.type(id);
+                Invariants.checkArgument(thisType.isSafeToMix(otherType), "Attempted to add %s; expected %s but was %s", id, thisType, otherType);
+            }
+            localTxns.addAll(other.localCommands);
+            txnToRange.putAll(other.commandsToRanges);
+            // If "put" was called before for a txn present in "other", to respect the "put" semantics that update must
+            // be removed from "rangeToTxn" (as it got removed from "txnToRange").
+            // The expected common case is that this method is called on an empty builder, so the removeIf is off an
+            // empty list (aka no-op)
+            rangeToTxn.removeIf(data -> other.commandsToRanges.containsKey(data.txnId));
+            rangeToTxn.addAll(other.rangesToCommands);
+            return (T) this;
+        }
+
+        public T mergeRemote(TxnId txnId, Ranges ranges, BiFunction<? super Ranges, ? super Ranges, ? extends Ranges> remappingFunction)
+        {
+            Invariants.checkArgument(!localTxns.contains(txnId), "Attempted to merge remote txn %s, but this is a local txn", txnId);
+            // accord.impl.CommandTimeseries.mapReduce does the check on status and deps type, and NotWitnessed should match the semantics hard coded in InMemorySafeStore...
+            // in that store, the remote history is only ever included when minStauts == null and deps == ANY... but mapReduce sees accord.local.Status.KnownDeps.hasProposedOrDecidedDeps == false
+            // as a mis-match, so will be excluded... since NotWitnessed will return false it will only be included IFF deps = ANY.
+            // When it comes to the minStatus check, the current usage is "null", "Committed", "Accepted"... so NotWitnessed will only be included in the null case;
+            // the only subtle difference is if minStatus = NotWitnessed, this API will include these but InMemoryStore won't
+            RangeCommandSummary oldValue = txnToRange.get(txnId);
+            RangeCommandSummary newValue = oldValue == null ?
+                                           new RangeCommandSummary(txnId, ranges, SaveStatus.NotWitnessed, null, Collections.emptyList())
+                                           : oldValue.withRanges(ranges, remappingFunction);
+            if (newValue == null)
+            {
+                remove(txnId);
+            }
+            else if (!oldValue.equalsDeep(newValue))
```

Comment:
When do we hit this case where oldValue is not null, but `newValue` has more information? Is it when we are a minority in some transaction that is missing information? Seems like if we have a local value we should never be loading remote info?

My only concern here is that this reverts our knowledge of the range transaction to a less knowledgable state in some situation.

Reply:
This is to keep the semantics of `Map.merge` which is used in in-memory store.  As far as I can tell, this happens when you cross epochs, as we collect `Deps` for a single epoch, so a different epoch may have more ranges; so we collect all

Reply:
Ah cool, forgot it merges them at the top of the function.